net/ipfrag/ipfrag.c

```diff
@@ -0,0 +1,1257 @@
+       * the non-zero fragments (copy to new I/O buffers), space for the
+       * L3 IP header must be reserved for all fragments
+       */
+
+      head = iob;
+      while (iob != NULL && nreside > iob->io_len)
+        {
+          nreside -= iob->io_len;
+          iob      = iob->io_flink;
+        }
+
+      leftstart = iob->io_data + iob->io_offset + nreside;
+      leftlen   = iob->io_len - nreside;
+
+      orig = iob->io_flink;
+      if (orig != NULL)
+        {
+          orig->io_pktlen = head->io_pktlen - (mtu + leftlen);
+          iob->io_flink   = NULL;
+        }
+
+      head->io_pktlen = mtu;
+      iob->io_len     = nreside;
+
+      if (iob_tryadd_queue(head, fragq) < 0)
+        {
+          goto allocfail;
+        }
+
+      head = NULL;
+      nfrags++;
+
+      if (leftlen == 0 && orig != NULL)
+        {
+          reorg = ip_fragout_allocfragbuf(fragq);
+          GOTO_IF(reorg == NULL, allocfail);
+
+          nfrags++;
+
+          /* This is a new fragment buffer, reserve L2&L3 header space
+           * in the front of this buffer
+           */
+
+          UPDATE_IOB(reorg, CONFIG_NET_LL_GUARDSIZE, unfraglen);
+        }
+      else
+        {
+          /* If the MTU is relatively small, the remaining data of the first
+           * I/O buffer may need to be fragmented multiple times.
+           * For IPv4, the first I/O Buffer is reused, which have reserved
+           * the L4 header space, the following fragmentation flow is only
+           * for non-zero fragments, so following flow does not need to
+           * consider the L4 header
+           */
+
+          while (leftlen > 0)
+            {
+              reorg = ip_fragout_allocfragbuf(fragq);
+              GOTO_IF(reorg == NULL, allocfail);
+
+              nfrags++;
+
+              if (leftlen + unfraglen > mtu)
+                {
+                  ncopy = mtu - unfraglen;
+                }
+              else
+                {
+                  ncopy = leftlen;
+                }
+
+              /* This is a new fragment buffer, reserve L2&L3 header space
+               * in the front of this buffer
+               */
+
+              UPDATE_IOB(reorg, CONFIG_NET_LL_GUARDSIZE, unfraglen);
+
+              /* Then copy L4 data */
+
+              GOTO_IF(iob_trycopyin(reorg, leftstart, ncopy,
+                    reorg->io_pktlen, false) < 0, allocfail);
+
+              leftstart     += ncopy;
+              leftlen       -= ncopy;
+            }
+        }
+    }
+#ifdef CONFIG_NET_IPv6
+  else
+#endif
+#endif
+#ifdef CONFIG_NET_IPv6
+  if (domain == PF_INET6)
+    {
+      unfraglen += EXTHDR_FRAG_LEN;
+
+      /* Fragmentation requires the length field to be a multiples of 8,
+       * and the length of the IPv6 basic header and all extended headers
+       * is a multiples of 8, so here directly fix the MTU to 8-byte
+       * alignment.
+       */
+
+      mtu = mtu >> 3 << 3;
```

Comment:
Ditto: Could it work replacing ">> 3 << 3" with "((mtu - IPv4_HDRLEN) & ~0x7)" ?